### ISSUE TITLE
channel last with checkpoint throws runtime error: set_sizes_and_strides is not allowed on a Tensor created from .data or .detach().

### ISSUE THREAD

**Issue by kingder** [pytorch/pytorch#48439]
## 🐛 Bug

runtime error when channel last and checkpoint used together.

## To Reproduce
```
import torch
from torch.utils.checkpoint import checkpoint

input = torch.randint(1, 10, (2, 8, 4, 4), dtype=torch.float32, device="cuda", requires_grad=True)
model = torch.nn.Conv2d(8, 4, 3).cuda().float()

input = input.contiguous(memory_format=torch.channels_last)
model = model.to(memory_format=torch.channels_last) # Module parameters need to be Channels Last

out = checkpoint(model, input).sum()

out.backward()
```
## Expected behavior

no errors should be thrown.

## Environment

PyTorch version: 1.7.0+cu110
Is debug build: True
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 5.4.0
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.6 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 

Nvidia driver version: 455.32.00
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip] numpy==1.17.3
[pip] torch==1.7.0+cu110
[pip] torchaudio==0.7.0
[pip] torchvision==0.8.1+cu110

**Comment by albanD** [comment]
I can repro on cuda. But the code runs just fine on CPU.

**Comment by albanD** [comment]
Simpler repro that does not involve checkpointing:
```
import torch
from torch.utils.checkpoint import checkpoint

input = torch.randint(1, 10, (2, 8, 4, 4), dtype=torch.float32, device="cuda", requires_grad=True)
model = torch.nn.Conv2d(8, 4, 3).cuda().float()

input = input.contiguous(memory_format=torch.channels_last)
model(input.detach())
```

This looks wrong as the input/weight should not be modified inplace: https://github.com/pytorch/pytorch/blob/b5149513ec02fa7d9711f6f79f759654f5d3a840/aten/src/ATen/native/cudnn/Conv.cpp#L977

**Comment by gchanan** [comment]
CC @VitalyFedyunin 

this seems to not match the documentation, which suggests `resize` on the same shape should be a no-op.

**Comment by kingder** [comment]
> Simpler repro that does not involve checkpointing:
> 
> ```
> import torch
> from torch.utils.checkpoint import checkpoint
> 
> input = torch.randint(1, 10, (2, 8, 4, 4), dtype=torch.float32, device="cuda", requires_grad=True)
> model = torch.nn.Conv2d(8, 4, 3).cuda().float()
> 
> input = input.contiguous(memory_format=torch.channels_last)
> model(input.detach())
> ```
> 
> This looks wrong as the input/weight should not be modified inplace:
> 
> https://github.com/pytorch/pytorch/blob/b5149513ec02fa7d9711f6f79f759654f5d3a840/aten/src/ATen/native/cudnn/Conv.cpp#L977

Is there any workaround?

**Comment by albanD** [comment]
Yes sorry, you just need to avoid having a conv as the first module in the checkpoint when using a different memory format.
The cheapest thing for you is to update your original code to do this for example (the view op does not do any copy so will be very cheap): `out = checkpoint(lambda inp: model(inp.view_as(inp)), input).sum()`

**Comment by VitalyFedyunin** [comment]
> CC @VitalyFedyunin
> 
> this seems to not match the documentation, which suggests `resize` on the same shape should be a no-op.

Yes, this looks like documentation issue and also changing input tensor (even metadata) inplace is shady. I will assign it to myself to followup with fixes. 

If @kingder are willing to contribute code, I'm ready to provide more detailed explanation how to fix conv operator.

**Comment by zixeljupo** [comment]
@VitalyFedyunin I want to contribute with this one, what do I need for that?

**Comment by robert780612** [comment]
I also suffer from this issue, and I found a workaround as the following.

```
import torch
from torch.utils.checkpoint import checkpoint

input = torch.randint(1, 10, (2, 8, 4, 4), dtype=torch.float32, device="cuda", requires_grad=True)
model = torch.nn.Conv2d(8, 4, 3).cuda().float()

with torch.no_grad():
    input = input.contiguous(memory_format=torch.channels_last)

model(input)
```

**Comment by albanD** [comment]
@zixeljupo You would need to update https://github.com/pytorch/pytorch/blob/b5149513ec02fa7d9711f6f79f759654f5d3a840/aten/src/ATen/native/cudnn/Conv.cpp#L977 to either make the resize out of place. Or first take an alias of the weights before doing the inplace resize. You should also add one of the repro from this issue as a test in test_nn.py to make sure this doesn't happen again in the future.

**Comment by zixeljupo** [comment]
> @zixeljupo You would need to update
> 
> https://github.com/pytorch/pytorch/blob/b5149513ec02fa7d9711f6f79f759654f5d3a840/aten/src/ATen/native/cudnn/Conv.cpp#L977
> 
> to either make the resize out of place. Or first take an alias of the weights before doing the inplace resize. You should also add one of the repro from this issue as a test in test_nn.py to make sure this doesn't happen again in the future.

Thank you! Despite I wrote own workaround in python, I'll try to fix this.
